src/cryptonote_core/blockchain.cpp

```diff
@@ -3405,6 +3405,113 @@ void Blockchain::flush_invalid_blocks() {
     m_invalid_blocks.clear();
 }
 //------------------------------------------------------------------
+std::vector<eth::bls_public_key> Blockchain::get_removable_nodes() const
+{
+    // TODO: Just calculate an acceleration structure on block receive.
+    std::vector<eth::bls_public_key> bls_pubkeys_in_snl;
+    std::vector<eth::bls_public_key> bls_pubkeys_in_smart_contract;
+    {
+        // NOTE: Extract all the service nodes from the Oxen work-chain _excluding_ those that have
+        // been recently removed (e.g. voluntarily exited or dereg by protocol). Recently removed
+        // nodes on the Oxen side need to be removed from the smart contract side and this and can
+        // be done by aggregating a signature (or without if the wait time has elapsed) from the
+        // network to be removed from the smart contract (which gets re-witnessed by Oxen thereby
+        // removing it from the 'recently_removed_nodes' list).
+        auto sns = service_node_list.get_service_node_list_state();
+        bls_pubkeys_in_snl.reserve(sns.size());
+        for (const auto& sni : sns)
+            bls_pubkeys_in_snl.push_back(sni.info->bls_public_key);
+
+        // NOTE: Because of vote confirmations, we have nodes that can exist in the smart contract
+        // but not the service node list for awhile until they get confirmed at which point they get
+        // added to the SNL.
+        //
+        // We assume that these nodes will be added to the SNL. If it does, it will eventually be
+        // removed from this list and show up in the 'get_service_node_list_state()' branch above.
+        //
+        // If the node gets denied, it will be removed from this list and never added to the SNL.
+        service_node_list.for_each_pending_l2_state(
+            [&bls_pubkeys_in_snl]<typename Event>(const Event& e, const service_nodes::service_node_list::unconfirmed_l2_tx&) {
+                if constexpr  (std::is_same_v<Event, eth::event::NewServiceNode>) {
+                    bls_pubkeys_in_snl.push_back(e.bls_pubkey);
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExitRequest>) {
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExit>) {
+                } else {
+                    log::error(logcat, "Got unknown event type when iterating each pending L2 state from the SNL!");
+                }
+        });
+
+        // NOTE: Extract all service nodes from the smart contract
+        eth::RewardsContract::ServiceNodeIDs smart_contract_ids =
+                m_l2_tracker->get_all_service_node_ids(std::nullopt);
+        bls_pubkeys_in_smart_contract = std::move(smart_contract_ids.bls_pubkeys);
```

Comment:
The member of an rvalue reference is itself an rvalue reference, which means you can change this as follows and preserve move semantics without needing a named variable + explicit std::move:
```suggestion
        bls_pubkeys_in_smart_contract = m_l2_tracker->get_all_service_node_ids(std::nullopt).bls_pubkeys;
```

(Feel free to ignore if you prefer the current two-step approach: I'd be surprised if the compiler doesn't optimize both things down to pretty much the same thing anyway).

Little example to demonstrate:
```C++
#include <string>
#include <iostream>

struct A { std::string s; };
struct B { A a; };

void f(const A& a) { std::cout << "f(const lvalue) called with " << a.s << "\n"; };
void f(A&& a) { std::cout << "f(rvalue) called with " << a.s << "\n"; };

B make_b(std::string s) {
    B result;
    result.a.s = std::move(s);
    return result;
}

int main() {
    B b1 = make_b("abc");
    f(b1.a);
    f(std::move(b1.a));
    f(std::move(b1).a);
    f(make_b("xyz").a);
}
```
```
f(const lvalue) called with abc
f(rvalue) called with abc
f(rvalue) called with abc
f(rvalue) called with xyz
```

Reply:
This is a good suggestion